Validate quantities are valid in the Add to Cart + Options on setQuantity()

Aljullu opened the pull request:
### Changes proposed in this Pull Request:

_(Issue originally highlighted by @kmanijak in https://github.com/woocommerce/woocommerce/pull/60543#pullrequestreview-3143424300)_

This PR implements some changes to how quantity validation is handled in the _Add to Cart + Options_ block:

* Now, we validate quantities inside `setQuantity()`, this has the advantage that the _Add to Cart_ button can change its styles from disabled/enabled instantly, instead of having to wait until the inputs are unfocused.
* When introducing a value below the minimum allowed quantity, we don't automatically update it to min. For example, if the minimum quantity accepted for a product is `4` and the shopper introduces `2`, we don't change it to `4`. That's to avoid the case that the shopper introduces `2` and, without unfocusing the input, clicks on the button, which would unexpectedly add 4 items. This is also aligned with how things behave in classic templates.

### How to test the changes in this Pull Request:

1. Add this code snippet to modify the `min`, `max` and `step` quantities of one of the simple products (you will need to change `110` with the product id):

```PHP
add_filter( 'woocommerce_quantity_input_min', function( $default, $product ) {
	if ( $product->get_id() === 110 ) {
		return 4;
	}
	return $default;
}, 10, 2 );

add_filter( 'woocommerce_quantity_input_step', function( $default, $product ) {
	if ( $product->get_id() === 110 ) {
		return 2;
	}
	return $default;
}, 10, 2 );

add_filter( 'woocommerce_quantity_input_max', function( $default, $product ) {
	if ( $product->get_id() === 110 ) {
		return 8;
	}
	return $default;
}, 10, 2 );
```
2. Go to _Appearance_ > _Editor_ > _Templates_ > _Single Product_.
3. Click on the _Add to Cart with Options_ block and update to the blockified version.
4. Go to the frontend of the simple product that you added the quantity constraints to. In another tab, go to the frontend of a grouped product containing the simple product with quantity constraints.
5. Verify that:
  * If you leave the input to an empty string, a space or a random string, it gets automatically updated to the min product value (or 0 in grouped product children).
  * If you set a numeric value smaller than min, it doesn't auto-update, but the _Add to Cart_ button looks disabled.
  * In simple products, you can't set the value to 0.
  * In grouped products, if all products are set to 0, the _Add to Cart_ button looks disabled. It only gets enabled if at least one of the products get a quantity value higher than its minimum.
  * You can still use the reduce and increase quantity buttons as usual.
  
Before | After
--- | ---
<img width="1515" height="976" alt="imatge" src="https://github.com/user-attachments/assets/8c0fc389-0501-473e-9499-af7e59289fb5" /> | <img width="1515" height="976" alt="imatge" src="https://github.com/user-attachments/assets/b925d3e0-be01-46f2-9d1c-6fdb36f60b8a" />

Aljullu committed:
Add to Cart + Options: validate quantities on input

Aljullu committed:
Add changelog file

Aljullu committed:
Simplify code

Aljullu committed:
Implement CodeRabbit suggestions and code cleanup

Aljullu committed:
Remove unnecessary checks because productData() always returns step, min and max

Aljullu committed:
Linting

Aljullu committed:
Make validateQuantity() argument optional

Aljullu committed:
Make getProductData() parameters consistent

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/frontend.ts:
```diff
@@ -298,6 +291,30 @@ const { actions, state } = store<
+						group: 'invalid-quantities',
+					} );
+				}
+			},
```
This should have been fixed in 46af7c2a4962722d0c114f1cfe1f0cf87ca6fce1 and 5dcbec6aef203811cf2789b92995a14786152f38.

gigitux commented:
I didn't check the code yet, but I want to confirm some behaviors:

> In simple products, you can't set the value to 0.


https://github.com/user-attachments/assets/7b8b7d39-f39d-42bb-92e8-060f718cf474

Is this the right behavior? 

> If you leave the input to an empty string, a space or a random string, it gets automatically updated to the min product value (or 0 in grouped product children).

In the case of a grouped product, the input isn't updated:


https://github.com/user-attachments/assets/512d6a8a-dbb3-4304-a4c7-92795ed88801

Aljullu committed:
Don't allow setting the input to 0 in simple products and make sure letters are reset

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/frontend.ts:
```diff
@@ -288,16 +277,46 @@ const { actions, state } = store<
 			},
 		},
 		actions: {
+			validateQuantity( value?: number ) {
+				actions.clearErrors( 'invalid-quantities' );
+
+				if ( typeof value !== 'number' ) {
+					return;
+				}
+
+				const context = getContext< Context >();
+
+				// If selected quantity is invalid, add an error.
+				const { variationId } = state;
+				const id = variationId || context.productId;
+				const productObject = getProductData(
+					id,
+					context.productType,
+					context.availableVariations,
+					context.selectedAttributes
+				);
+
+				if (
+					value === 0 ||
+					( productObject &&
+						( value < productObject.min ||
+							value > productObject.max ) )
+				) {
+					const { errorMessages } = getConfig();
+
+					actions.addError( {
+						code: 'invalidQuantities',
+						message: errorMessages?.invalidQuantities || '',
+						group: 'invalid-quantities',
+					} );
+				}
+			},
```
`validateQuantity()` is overridden in the _Grouped Product Selector_ store, so `validateQuantity()` will never go through this code path when it's a grouped product child.

gigitux replied:
I would add a comment explaining that `validateQuantity` is overridden in the Grouped Product Selector store. No strong opinion btw.

Aljullu commented:
Ups, that wasn't right. Thanks for noticing, @gigitux! Apparently there are some oddities in the `onchange` event when a numeric input is filled with text or empty strings if the input was already invalid. I guess I originally only tested writing letters _after_ having already written a number, my bad. :sweat_smile:

In any case, I switched to using the `blur` event and refactored the logic a bit (TIL about [`valueAsNumber`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/valueAsNumber)). I also added some extra tests to cover these cases.

gigitux reviewed (approved):
Thanks for addressing the feedback! Everything works as expected:

### Simple Product

https://github.com/user-attachments/assets/a1afec56-ef62-4e98-89a0-80161df31fd1

### Grouped Product

https://github.com/user-attachments/assets/06f33e94-40ec-44b0-a6fc-f520da875e61
